iothub/service/src/Messaging/FileUploadNotificationProcessorClient.cs

```diff
@@ -0,0 +1,215 @@
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Func<FileNotification, DeliveryAcknowledgement> _fileNotificationProcessor;
+
+        /// <summary>
+        /// The callback to be executed when the connection is lost.
+        /// </summary>
+        public Action<ErrorContext> _errorProcessor;
+
+        /// <summary>
+        /// Creates an instance of this class. Provided for unit testing purposes only.
+        /// </summary>
+        protected FileUploadNotificationProcessorClient()
+        {
+        }
+
+        internal FileUploadNotificationProcessorClient(
+            string hostName,
+            IotHubConnectionProperties credentialProvider,
+            IotHubServiceClientOptions options)
+        {
+            _hostName = hostName;
+            _credentialProvider = credentialProvider;
+            _connection = new IotHubConnection(credentialProvider, options.UseWebSocketOnly, options);
+            FileNotificationReceiver = new AmqpFileNotificationReceiver(_connection);
+        }
+
+        /// <summary>
+        /// Gets the AmqpFileNotificationReceiver which receives notifications for file upload operations.
+        /// </summary>
+        internal AmqpFileNotificationReceiver FileNotificationReceiver;
+
+        /// <summary>
+        /// Open the connection and start receiving file upload notifications.
+        /// </summary>
+        /// <exception cref="IotHubCommunicationException">Thrown if the client encounters a transient retriable exception. </exception>
+        /// <exception cref="IotHubCommunicationException">Thrown when the operation has been canceled. The inner exception will be
+        /// <see cref="OperationCanceledException"/>.</exception>
+        /// <exception cref="SocketException">Thrown if a socket error occurs.</exception>
+        /// <exception cref="WebSocketException">Thrown if an error occurs when performing an operation on a WebSocket connection.</exception>
+        /// <exception cref="IOException">Thrown if an I/O error occurs.</exception>
+        /// <exception cref="IotHubException">Thrown if an error occurs when communicating with IoT hub service.
+        /// If <see cref="IotHubException.IsTransient"/> is set to <c>true</c> then it is a transient exception.
+        /// If <see cref="IotHubException.IsTransient"/> is set to <c>false</c> then it is a non-transient exception.</exception>
+        public virtual async Task OpenAsync()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Opening FileUploadNotificationProcessorClient", nameof(OpenAsync));
+
+            try
+            {
+                if (_fileNotificationProcessor == null)
+                {
+                    throw new Exception("Callback for message feedback {0} must be set before opening the connection.".FormatInvariant(_fileNotificationProcessor));
```

Comment:
We'll likely change this exception type later with Bryce's work, so it's fine to leave this as Exception for now. I don't see any benefit to adding the null file upload notification processor to the message, though.

```suggestion
                    throw new Exception("Callback for file upload notifications must be set before opening the connection.");
```